File: test/ingest/claim_mapping_test.exs
# test/ingest/claim_mapping_test.exs — ExUnit for the claim mapper (bead gr-beo).
#
#   Run:  mix test
#
# Covers the fold semantics (set/add/remove/delete/clear), canonicalize+partition (shared set),
# and the claim shapes — against synthetic envelopes AND the real 422156 fixture, driven all the
# way through the engine (cluster + reconcile) to prove the loop end-to-end. Modules compiled
# from lib/; ExUnit starts in test/test_helper.exs.

defmodule ClaimMappingTest do
  use ExUnit.Case, async: true

  @fixture Path.join(__DIR__, "fixtures/medipim_be_422156.json")

  # ── helpers ──────────────────────────────────────────────────────────────
  defp envelope(entity, events) do
    {:ok, env} =
      HistoryEnvelope.from_map(%{
        "schema_version" => "1",
        "legacy_entity" => entity,
        "events" => events
      })

    env
  end

  defp id(source, op, scheme, code, at),
    do: %{
      "recorded_at" => at,
      "source" => source,
      "op" => op,
      "kind" => "identity",
      "scheme" => scheme,
      "code" => code
    }

  defp clusters(%{claims: claims, shared: shared}),
    do: Cluster.variants(Substrate.current(claims), shared)

  # ── fold semantics ─────────────────────────────────────────────────────────
  describe "fold" do
    test "set replaces a single-valued scheme" do
      env = envelope(1, [id("A", "set", "cnk", "111", 10), id("A", "set", "cnk", "222", 20)])
      assert ClaimMapping.listings([env]) == %{{1, "A"} => MapSet.new([{:cnk, "222"}])}
    end

    test "add accumulates, remove deletes one" do
      env =
        envelope(1, [
          id("A", "add", "ean", "5012345678900", 10),
          id("A", "add", "ean", "4006381333931", 20),
          id("A", "remove", "ean", "5012345678900", 30)
        ])

      assert ClaimMapping.listings([env]) == %{{1, "A"} => MapSet.new([{:gtin, "04006381333931"}])}
    end

    test "delete (op-4) drops the whole scheme entry" do
      env =
        envelope(1, [
          id("A", "set", "eanGtin13", "5012345678900", 10),
          id("A", "delete", "eanGtin13", "A", 20)
        ])

      assert ClaimMapping.listings([env]) == %{}
    end

    test "set-null clears the code" do
      env =
        envelope(1, [id("A", "set", "eanGtin14", "05012345678900", 10), id("A", "set", "eanGtin14", nil, 20)])

      assert ClaimMapping.listings([env]) == %{}
    end

    test "canonicalize merges an EAN-13 and its GTIN-14 form into one code" do
      env =
        envelope(1, [
          id("A", "add", "ean", "5012345678900", 10),
          id("A", "set", "eanGtin14", "05012345678900", 20)
        ])

      assert ClaimMapping.listings([env]) == %{{1, "A"} => MapSet.new([{:gtin, "05012345678900"}])}
    end
  end

  # ── partition / shared ──────────────────────────────────────────────────────
  describe "partition" do
    test "a restricted (in-store) GTIN lands in the shared set" do
      env = envelope(1, [id("A", "add", "gtin", "02000000000000", 10), id("A", "set", "cnk", "111", 20)])
      %{shared: shared} = ClaimMapping.build([env])
      assert shared == MapSet.new([{:gtin, "02000000000000"}])
    end

    test "bridging codes (cnk, normal gtin) are not shared" do
      env = envelope(1, [id("A", "set", "cnk", "111", 10), id("A", "add", "gtin", "5012345678900", 20)])
      assert ClaimMapping.build([env]).shared == MapSet.new()
    end
  end

  # ── claim shapes ─────────────────────────────────────────────────────────────
  describe "claims" do
    test "one identity claim per listing, carrying the folded code-set" do
      env = envelope(7, [id("A", "set", "cnk", "111", 10), id("B", "set", "cnk", "222", 10)])
      ids = Enum.filter(ClaimMapping.build([env]).claims, &(&1.kind == :identity))
      assert length(ids) == 2
      refs = ids |> Enum.map(& &1.data.ref) |> Enum.sort()
      assert refs == ["7:A", "7:B"]
    end

    test "attribute is anchored to the listing's primary code (CNK ▸ GTIN)" do
      env =
        envelope(1, [
          id("A", "set", "cnk", "111", 10),
          id("A", "add", "gtin", "5012345678900", 10),
          %{
            "recorded_at" => 20,
            "source" => "A",
            "op" => "set",
            "kind" => "attribute",
            "field" => "name",
            "locale" => "fr",
            "value" => "Crème"
          }
        ])

      attr = Enum.find(ClaimMapping.build([env]).claims, &(&1.kind == :attribute))
      assert attr.data.code == {:cnk, "111"}
      assert attr.data.field == "name:fr"
      assert attr.data.value == "Crème"
    end

    test "grouping points every code at the legacy entity" do
      env = envelope(99, [id("A", "set", "cnk", "111", 10)])
      grp = Enum.find(ClaimMapping.build([env]).claims, &(&1.kind == :grouping))
      assert grp.data == %{code: {:cnk, "111"}, product: 99}
    end
  end

  # ── cross-entity behaviour through the engine ───────────────────────────────
  describe "through the engine" do
    test "two entities sharing a CNK cluster into one key, flagged as a collision" do
      envs = [
        envelope(1, [id("A", "set", "cnk", "100", 10)]),
        envelope(2, [id("B", "set", "cnk", "100", 10)])
      ]

      %{claims: claims, shared: shared} = ClaimMapping.build(envs)
      live = Substrate.current(claims)

      assert [_one_cluster] = Cluster.variants(live, shared)

      ledger =
        IdentityLedger.decide(IdentityLedger.new(), {:reconcile, Cluster.variants(live, shared), 10})
        |> Enum.reduce(IdentityLedger.new(), &IdentityLedger.evolve(&2, &1))

      collisions = Stewardship.detect_collisions(ledger.members, live, 10)
      assert [%Events.ConflictFlagged{subject: {:collision, _key}}] = collisions
    end
  end

  # ── the real 422156 fixture ─────────────────────────────────────────────────
  describe "real entity 422156" do
    setup do
      {:ok, env} = HistoryEnvelope.load(@fixture)
      %{env: env}
    end

    test "org 44 converged: dropped its old EAN, holds CNK + canonical GTIN", %{env: env} do
      l = ClaimMapping.listings([env])
      assert l[{422_156, "44"}] == MapSet.new([{:cnk, "3612173"}, {:gtin, "03282770146004"}])
      # the 2018 EAN it later removed must be gone
      refute MapSet.member?(l[{422_156, "44"}], {:gtin, "03282770049374"})
    end

    test "org 1035 kept its extra (never-removed) GTIN", %{env: env} do
      l = ClaimMapping.listings([env])
      assert MapSet.member?(l[{422_156, "1035"}], {:gtin, "03282770114577"})
      assert MapSet.member?(l[{422_156, "1035"}], {:cnk, "3612173"})
    end

    test "all three listings collapse to ONE surrogate key (legacy was right)", %{env: env} do
      result = ClaimMapping.build([env])
      assert result.shared == MapSet.new()

      assert [cluster] = clusters(result)
      assert MapSet.member?(cluster, {:cnk, "3612173"})
      assert MapSet.member?(cluster, {:gtin, "03282770146004"})
      assert MapSet.member?(cluster, {:gtin, "03282770114577"})

      minted =
        IdentityLedger.decide(IdentityLedger.new(), {:reconcile, clusters(result), 1})
        |> Enum.filter(&match?(%Events.IdentityMinted{}, &1))

      assert length(minted) == 1
    end
  end
end

File: php/tests/ClaimMappingTest.php
<?php

declare(strict_types=1);

namespace GoldenRecord\Tests;

use GoldenRecord\ClaimMapping;
use GoldenRecord\Cluster;
use GoldenRecord\EnvelopeLoader;
use GoldenRecord\Lanes;
use GoldenRecord\Sets;
use GoldenRecord\Substrate;
use PHPUnit\Framework\TestCase;

/**
 * Ported from test/ingest/claim_mapping_test.exs: the fold semantics (set/add/remove/delete/clear),
 * canonicalize+partition (shared set), claim shapes, and the real 422156 convergence.
 */
final class ClaimMappingTest extends TestCase
{
    private const FIXTURE = __DIR__.'/../../test/ingest/fixtures/medipim_be_422156.json';

    private function envelope(int $entity, array $events): array
    {
        [$ok, $env] = EnvelopeLoader::fromMap(['schema_version' => '1', 'legacy_entity' => $entity, 'events' => $events]);
        self::assertSame('ok', $ok);

        return $env;
    }

    private function id(string $source, string $op, string $scheme, ?string $code, int $at): array
    {
        return ['recorded_at' => $at, 'source' => $source, 'op' => $op, 'kind' => 'identity', 'scheme' => $scheme, 'code' => $code];
    }

    /** @return list<array{0: string, 1: string}> the listing's codes, sorted */
    private function listingCodes(array $envelopes, int $entity, string $source): array
    {
        $listings = ClaimMapping::listings($envelopes);
        $key = "i:$entity\x1f$source";

        return isset($listings[$key]) ? Sets::valuesSorted($listings[$key]) : [];
    }

    public function test_set_replaces_a_single_valued_scheme(): void
    {
        $env = $this->envelope(1, [$this->id('A', 'set', 'cnk', '111', 10), $this->id('A', 'set', 'cnk', '222', 20)]);
        self::assertSame([['cnk', '222']], $this->listingCodes([$env], 1, 'A'));
    }

    public function test_add_accumulates_remove_deletes_one(): void
    {
        $env = $this->envelope(1, [
            $this->id('A', 'add', 'ean', '5012345678900', 10),
            $this->id('A', 'add', 'ean', '4006381333931', 20),
            $this->id('A', 'remove', 'ean', '5012345678900', 30),
        ]);
        self::assertSame([['gtin', '04006381333931']], $this->listingCodes([$env], 1, 'A'));
    }

    public function test_delete_drops_the_whole_scheme(): void
    {
        $env = $this->envelope(1, [
            $this->id('A', 'set', 'eanGtin13', '5012345678900', 10),
            $this->id('A', 'delete', 'eanGtin13', 'A', 20),
        ]);
        self::assertSame([], ClaimMapping::listings([$env]));
    }

    public function test_set_null_clears(): void
    {
        $env = $this->envelope(1, [
            $this->id('A', 'set', 'eanGtin14', '05012345678900', 10),
            $this->id('A', 'set', 'eanGtin14', null, 20),
        ]);
        self::assertSame([], ClaimMapping::listings([$env]));
    }

    public function test_unrecognised_scheme_stays_a_string(): void
    {
        $env = $this->envelope(1, [$this->id('A', 'set', 'mysteryScheme', 'XYZ', 10)]);
        self::assertSame([['mysteryScheme', 'XYZ']], $this->listingCodes([$env], 1, 'A'));
    }

    public function test_french_fields_map_to_their_schemes(): void
    {
        $env = $this->envelope(1, [
            $this->id('A', 'set', 'cipOrAcl7', '4440813', 10),
            $this->id('A', 'set', 'acl13', '3401344408137', 20),
        ]);
        self::assertSame([['acl13', '3401344408137'], ['cip_acl7', '4440813']], $this->listingCodes([$env], 1, 'A'));
    }

    public function test_restricted_gtin_lands_in_shared(): void
    {
        $env = $this->envelope(1, [$this->id('A', 'add', 'gtin', '02000000000000', 10), $this->id('A', 'set', 'cnk', '111', 20)]);
        $built = ClaimMapping::build([$env]);
        self::assertSame([['gtin', '02000000000000']], Sets::values($built['shared']));
    }

    public function test_bridging_codes_not_shared(): void
    {
        $env = $this->envelope(1, [$this->id('A', 'set', 'cnk', '111', 10), $this->id('A', 'add', 'gtin', '5012345678900', 20)]);
        self::assertSame([], ClaimMapping::build([$env])['shared']);
    }

    public function test_one_identity_claim_per_listing(): void
    {
        $env = $this->envelope(7, [$this->id('A', 'set', 'cnk', '111', 10), $this->id('B', 'set', 'cnk', '222', 10)]);
        $ids = array_filter(ClaimMapping::build([$env])['claims'], static fn ($c): bool => $c['kind'] === 'identity');
        self::assertCount(2, $ids);
        $refs = array_map(static fn ($c): string => $c['data']['ref'], $ids);
        sort($refs);
        self::assertSame(['7:A', '7:B'], $refs);
    }

    public function test_attribute_anchored_to_primary_cnk(): void
    {
        $env = $this->envelope(1, [
            $this->id('A', 'set', 'cnk', '111', 10),
            $this->id('A', 'add', 'gtin', '5012345678900', 10),
            ['recorded_at' => 20, 'source' => 'A', 'op' => 'set', 'kind' => 'attribute', 'field' => 'name', 'locale' => 'fr', 'value' => 'Crème'],
        ]);
        $attr = null;
        foreach (ClaimMapping::build([$env])['claims'] as $c) {
            if ($c['kind'] === 'attribute') {
                $attr = $c;
                break;
            }
        }
        self::assertSame(['cnk', '111'], $attr['data']['code']);
        self::assertSame('name:fr', $attr['data']['field']);
        self::assertSame('Crème', $attr['data']['value']);
    }

    // ── the real 422156 fixture ──────────────────────────────────────────────────

    public function test_org_44_converged_dropped_old_ean(): void
    {
        $env = EnvelopeLoader::loadBang(self::FIXTURE);
        $codes = $this->listingCodes([$env], 422156, '44');
        self::assertContains(['cnk', '3612173'], $codes);
        self::assertContains(['gtin', '03282770146004'], $codes);
        self::assertNotContains(['gtin', '03282770049374'], $codes);
    }

    public function test_all_listings_collapse_to_one_key(): void
    {
        $env = EnvelopeLoader::loadBang(self::FIXTURE);
        $result = ClaimMapping::build([$env]);
        self::assertSame([], $result['shared']);

        $clusters = Cluster::variants(Lanes::identityClaims(Substrate::current($result['claims']), 'product'), $result['shared']);
        self::assertCount(1, $clusters);
        self::assertTrue(Sets::member($clusters[0], ['cnk', '3612173']));
        self::assertTrue(Sets::member($clusters[0], ['gtin', '03282770146004']));
        self::assertTrue(Sets::member($clusters[0], ['gtin', '03282770114577']));
    }
}
